code/modules/reagents/chemistry/holder.dm

```diff
@@ -1,89 +1,225 @@
+	reaction_multiplier_overrun = multiplier % 1		//store current overrun
+	multiplier -= reaction_multiplier_overrun		//cut overrun
+
+	for(var/i in 1 to multiplier)
+		if(!react(++current_reaction_cycle))			//If returning false that means we've processed without a reaction happening, meaning we should be done.
+			stop_reacting()
+			break
+
+/datum/reagents/proc/react(cycle)
+	if(reagents_holder_flags & REAGENT_NOREACT)
+		return //Yup, no reactions here. No siree.
+
+	//Cache for hyper speed!
+	var/list/cached_reagents = reagent_list
+	var/list/cached_reactions = SSreagents.reactions_by_reagent_id
+	var/datum/cached_my_atom = my_atom
+
+	//Find what reactions we can do at the time of calculation
+	var/list/possible_reactions = get_possible_reactions()
+
+	//Do reactions
+	var/reaction_occurred = FALSE
+	var/continue_processing = FALSE
+	do
+		reaction_occurred = FALSE
+		if(possible_reactions.len)
+			var/datum/chemical_reaction/selected_reaction = possible_reactions[1]
+
+			//select the reaction with the most extreme temperature requirements
+			for(var/V in possible_reactions)
+				var/datum/chemical_reaction/competitor = V
+				if(selected_reaction.is_cold_recipe) //if there are no recipe conflicts, everything in possible_reactions will have this same value for is_cold_reaction. warranty void if assumption not met.
+					if(competitor.required_temp < selected_reaction.required_temp)
+						selected_reaction = competitor
+				else
+					if(competitor.required_temp > selected_reaction.required_temp)
+						selected_reaction = competitor
+
+			//Cache reaction variables
+			var/list/cached_required_reagents = selected_reaction.required_reagents
+			var/list/cached_results = selected_reaction.results
+
+			//How many times the formula can be applied to us
+			var/multiplier = INFINITY
+
+			//Limit by available reactants
+			for(var/B in cached_required_reagents)
+				multiplier = min(multiplier, get_reagent_amount(B) / cached_required_reagents[B])
+
+			//Limit by rate
+			multiplier = min(multiplier, selected_reaction.get_reaction_rate())
+
+			//React!
+			for(var/B in cached_required_reagents)
+				remove_reagent(B, (multiplier * cached_required_reagents[B]), safety = 1)
+
+			for(var/P in selected_reaction.results)
+				SSblackbox.record_feedback("tally", "chemical_reaction", cached_results[P]*multiplier, P)
+				add_reagent(P, cached_results[P] * multiplier, null, chem_temp)
+
+			WIP_TAG		//Find a better way to do this
+
+			if(cycle == 1)
+				var/list/seen = viewers(4, get_turf(my_atom))
+				var/iconhtml = icon2html(cached_my_atom, seen)
+				if(cached_my_atom)
+					if(!ismob(cached_my_atom)) // No bubbling mobs
+						if(selected_reaction.mix_sound)
+							playsound(get_turf(cached_my_atom), selected_reaction.mix_sound, 80, 1)
+
+						for(var/mob/M in seen)
+							to_chat(M, "<span class='notice'>[iconhtml] [selected_reaction.mix_message]</span>")
+
+					if(istype(cached_my_atom, /obj/item/slime_extract))
```

Comment:
Why is this hard coded instead of being a comsig thing

Reply:
remind me to add hooks for starting reactions later.